mcstatus/querier.py

```diff
@@ -128,6 +128,22 @@ def from_connection(cls, response: Connection):
         data = {}
         players = []
 
+        # If hostname is set to unicode, using other parameters of read_ascii() may be in the wrong order
+        key = response.received[:8].decode('ISO-8859-1')
```

Comment:
Do you have an example response that you could share? Preferably in a bytearray form that I could import. I'm curious why 8 bytes? Did you know that read_ascii already does the ISO-8859-1 decoding?

Reply:
> 

8 bytes because the ascii length of `gametype` is 8 bytes. The `hostname` parameter and value are followed by the `gametype` parameter  in the minecraft query protocol. 

read_ascii is not used here because only peek is needed, but the buffer does not need to be freed.

Response Example(python3.8 paper1.16.5 bytearray) :

bytearray(b'splitnum\x00\x80\x00hostname\x00111111111111111\x00gametype\x00SMP\x00game_id\x00MINECRAFT\x00version\x001.16.5\x00plugins\x00Paper on 1.16.5-R0.1-SNAPSHOT\x00map\x00world\x00numplayers\x000\x00maxplayers\x0020\x00hostport\x0025565\x00hostip\x00127.0.1.1\x00\x00\x01player_\x00\x00\x00')


As you can see in the above example, the `hostname` parameter is followed by the gametype

Bad Response Example(python3.8 paper1.16.5 bytearray) :

bytearray(b'splitnum\x00\x80\x00hostname\x00\x00*K\xd5\x00gametype\x00SMP\x00game_id\x00MINECRAFT\x00version\x001.16.5\x00plugins\x00Paper on 1.16.5-R0.1-SNAPSHOT\x00map\x00world\x00numplayers\x000\x00maxplayers\x0020\x00hostport\x0025565\x00hostip\x00127.0.1.1\x00\x00\x01player_\x00\x00\x00')

the hostname parameter has many \x00 , will caused parse wrong order.

Maybe my fix code looks weird, but he is correct, I have tested it on multiple versions of minecraft server and have used it in production.

Reply:
To summarize some of the Discord chat around this, I composed two test cases that this PR should address, but they are failing:
```python
    def test_query_handles_unicode_motd_with_nulls(self):
        self.querier.connection.receive(
            bytearray(
                b"splitnum\x00\x80\x00hostname\x00\x00*K\xd5\x00gametype\x00SMP\x00game_id\x00MINECRAFT\x00version\x001.16.5\x00plugins\x00Paper on 1.16.5-R0.1-SNAPSHOT\x00map\x00world\x00numplayers\x000\x00maxplayers\x0020\x00hostport\x0025565\x00hostip\x00127.0.1.1\x00\x00\x01player_\x00\x00\x00"
            )
        )
        response = self.querier.read_query()
        conn_bytes = self.querier.connection.flush()

    def test_query_handles_unicode_motd_with_2a00_at_the_start(self):
        self.querier.connection.receive(
            bytearray.fromhex(
                "73706c69746e756d008000686f73746e616d6500006f746865720067616d657479706500534d500067616d655f6964004d494e4543524146540076657273696f6e00312e31382e3100706c7567696e7300006d617000776f726c64006e756d706c61796572730030006d6178706c617965727300323000686f7374706f727400323535363500686f73746970003137322e31372e302e32000001706c617965725f000000"
            )
        )
        response = self.querier.read_query()
        conn_bytes = self.querier.connection.flush()
```
Both of these still need assertions, but they currently already hit an infinite loop or crash. We'll have to fix these to be passing with some logical assertions before this PR can be merged.

Reply:
Today I debug the test case you gave and know why there is a problem with the test case, you can try my modified test case.

    def test_query_handles_unicode_motd_with_nulls(self):
        self.querier.connection.receive(
            bytearray(
                b"\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00hostname\x00\x00*K\xd5\x00gametype\x00SMP\x00game_id\x00MINECRAFT\x00version\x001.16.5\x00plugins\x00Paper on 1.16.5-R0.1-SNAPSHOT\x00map\x00world\x00numplayers\x000\x00maxplayers\x0020\x00hostport\x0025565\x00hostip\x00127.0.1.1\x00\x00\x01player_\x00\x00\x00"
            )
        )
        response = self.querier.read_query()
        conn_bytes = self.querier.connection.flush()

    def test_query_handles_unicode_motd_with_2a00_at_the_start(self):
        self.querier.connection.receive(
            bytearray.fromhex(
                "00000000000000000000000000000000686f73746e616d6500006f746865720067616d657479706500534d500067616d655f6964004d494e4543524146540076657273696f6e00312e31382e3100706c7567696e7300006d617000776f726c64006e756d706c61796572730030006d6178706c617965727300323000686f7374706f727400323535363500686f73746970003137322e31372e302e32000001706c617965725f000000"
            )
        )
        response = self.querier.read_query()
        conn_bytes = self.querier.connection.flush()


I added some header data of the protocol, why do this, you can see `querier.py::read_query` and `querier.py::_read_packet`, and the previous `TestMinecraftQuerier::test_query` is also a data packet according to this format.

Reply:
> why do this
Because adding missing tests are important when fixing bugs so that the bugs do not come back later. With the different input in these tests (compared to `TestMinecraftQuerier::test_query`), it is triggering the bug and showcasing that it is not yet fixed.